pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -740,6 +863,106 @@ pub mod pallet {
 			};
 		}
 
+		pub(crate) fn calculate_collect(
+			loc: TrancheLocator<T::PoolId, T::TrancheId>,
+			order: UserOrder<T::Balance, T::EpochId>,
+			pool: PoolDetails<T::AccountId, T::CurrencyId, T::EpochId, T::Balance>,
+			end_epoch: T::EpochId,
+		) -> OutstandingCollections<T::Balance> {
+			let mut epoch_idx = order.epoch;
+
+			// No collect possible in this epoch
+			if epoch_idx == pool.current_epoch {
+				return OutstandingCollections {
+					payout_currency_amount: Zero::zero(),
+					payout_token_amount: Zero::zero(),
+					remaining_supply_currency: order.supply,
+					remaining_redeem_token: order.redeem,
+				};
+			}
+
+			// It is only possible to collect epochs which are already over
+			let parse_until_epoch = if end_epoch > pool.last_epoch_executed {
+				pool.last_epoch_executed
+			} else {
+				end_epoch
+			};
+
+			let mut payout_currency_amount: T::Balance = Zero::zero();
+			let mut payout_token_amount: T::Balance = Zero::zero();
+			let mut remaining_supply_currency = order.supply;
+			let mut remaining_redeem_token = order.redeem;
+			while epoch_idx <= parse_until_epoch
+				&& (remaining_supply_currency != Zero::zero()
+					|| remaining_redeem_token != Zero::zero())
+			{
+				if remaining_supply_currency != Zero::zero() {
+					let epoch = Epoch::<T>::try_get(&loc, epoch_idx)
+						.map_err(|_| Error::<T>::NoSuchPool)
+						.unwrap();
+
+					// TODO: this should round down, in favor of the system
+					let amount = T::BalanceRatio::checked_from_rational(
+						epoch.supply_fulfillment.deconstruct(),
+						Perquintill::ACCURACY,
+					)
+					.unwrap()
+					.checked_mul_int(remaining_supply_currency)
+					.unwrap();
```

Comment:
Is it safe to `unwrap()` here or are we expecting some errors? I would like to remove all unwraps and rather return errors.